packages/controls/src/controls/group/group.ts

```diff
@@ -0,0 +1,310 @@
+>
+
+type PropsData<C extends Config> = {
+  [K in keyof C['props']]?: DataType_<C['props'][K]>
+}
+
+type VersionedData<C extends Config> = {
+  [ControlDataTypeKey]:
+    | typeof Definition.v1DataType
+    | typeof ShapeV2Definition.v1DataType
+  value: PropsData<C>
+}
+
+type DataType<C extends Config> = ShapeData<C> | VersionedData<C>
+
+type ValueType<C extends Config> = {
+  [K in keyof C['props']]?: ValueType_<C['props'][K]>
+}
+
+type ResolvedValueType<C extends Config> = {
+  [K in keyof C['props']]: ResolvedValueType_<C['props'][K]>
+}
+type InstanceType<C extends Config> = GroupControl<Definition<C>>
+
+class Definition<C extends Config> extends ControlDefinition<
+  typeof Definition.type,
+  C,
+  DataType<C>,
+  ValueType<C>,
+  ResolvedValueType<C>,
+  InstanceType<C>
+> {
+  static readonly type = 'makeswift::controls::group' as const
+  static readonly v1DataType = 'group::v1' as const
+
+  static readonly Layout = {
+    Inline: `${this.type}::layout::inline`,
+    Popover: `${this.type}::layout::popover`,
+  } as const
+
+  static deserialize(
+    data: DeserializedRecord,
+    deserializeCallback: (r: DeserializedRecord) => ControlDefinition,
+  ): GroupDefinition {
+    if (data.type !== Definition.type)
+      throw new Error(
+        `Group: expected '${Definition.type}', got '${data.type}'`,
+      )
+
+    const {
+      config: { props, ...config },
+    } = z
+      .object({
+        type: z.literal(Definition.type),
+        config: z.object({
+          label: z.string().optional(),
+          preferredLayout: z
+            .union([
+              z.literal(Definition.Layout.Inline),
+              z.literal(Definition.Layout.Popover),
+            ])
+            .optional(),
+          props: z.record(z.string(), SerializationSchema.deserializedRecord),
+        }),
+      })
+      .parse(data)
+
+    const deserializedProps = mapValues(props, (itemDef) => {
+      return deserializeCallback(itemDef)
+    })
+
+    return Group({
+      props: deserializedProps,
+      ...config,
+    })
+  }
+
+  constructor(
+    config: C,
+    readonly dataType:
+      | typeof Definition.v1DataType
+      | typeof ShapeV2Definition.v1DataType,
+  ) {
+    super(config)
+  }
+
+  private get dataSignature() {
+    return { [ControlDataTypeKey]: this.dataType }
+  }
+
+  get propDefs() {
+    return this.config.props
+  }
+
+  get controlType() {
+    return Definition.type
+  }
+
+  get schema() {
+    const version = z.literal(1)
+    const type = z.literal(Definition.type)
+
+    const keys = mapValues(this.propDefs, (def) => def.schema)
+
+    const propsData = z.object(
+      mapValues(keys, (key) => key.data),
+    ) as SchemaType<PropsData<C>>
+
+    const versionedData = z.object({
+      [ControlDataTypeKey]: z.union([
+        z.literal(Definition.v1DataType),
+        z.literal(ShapeV2Definition.v1DataType),
+      ]),
+      value: propsData,
+    }) as SchemaType<VersionedData<C>>
```

Comment:
Note that the runtime schema allows `Group` to parse/validate all forms of data, regardless of the data type it's been initialized with.